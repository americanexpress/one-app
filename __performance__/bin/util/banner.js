/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const table = require('markdown-table');
const { Table } = require('console-table-printer');
const {
  testType, time, timeDiff, colors,
} = require('./format');
const { sum } = require('./math');

const metrics = {
  ...require('../k6/metrics'),
  ...require('../prometheus/metrics'),
};

function labelledMetric(metric, data, processor) {
  const { bold } = colors();
  return `${metrics[metric].label}: ${bold(metrics[metric].format(processor(data[metric])))}`;
}

function banner(colData, markdown = false) {
  const rows = [
    colData.map(({ meta }) => testType(meta.type)),
    colData.map(({ meta }) => time(meta.timestamp, 'Completed:')),
    colData.map(({ meta }) => timeDiff(meta.start, meta.end, 'Duration')),
    colData.map(({ data }) => labelledMetric('http_reqs', data, sum)),
    colData.map(({ data }) => labelledMetric('data_received', data, sum)),
    colData.map(({ data }) => labelledMetric('http_req_failed', data, sum)),
  ];

  if (colData.some(({ id }) => !!id)) rows.push(colData.map(({ id }) => id));

  if (markdown) {
    return table([
      colData.map(({ label }) => label),
      ...rows,
    ]);
  }

  const consoleTable = new Table({
    columns: colData.map(({ label }) => ({ name: label, alignment: 'left', color: 'none' })),
    colorMap: { none: '' },
  });

  consoleTable.addRows(rows.map((row) => row.reduce((acc, curr, i) => ({
    ...acc, [colData[i].label]: curr,
  }), {})));

  return consoleTable.render();
}

module.exports = banner;
