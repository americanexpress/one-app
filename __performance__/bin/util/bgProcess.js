/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs');
const { join } = require('node:path');
const { tmpdir } = require('node:os');
const pkg = require('../../../package.json');

const dir = join(tmpdir(), `one-perf-${pkg.version}`);
const filePath = join(dir, 'processes.json');

fs.mkdirSync(dir, { recursive: true });

const getLogFile = (name) => join(dir, `${name}.log`);

async function saveProcess(name, pid) {
  let processes = {};
  if (fs.existsSync(filePath)) {
    try {
      processes = JSON.parse(fs.promises.readFile(filePath));
    } catch (error) { /* ignore */ }
  }
  processes[name] = pid;
  await fs.promises.writeFile(filePath, JSON.stringify(processes, null, 2));
}

async function killProcess(name) {
  if (!fs.existsSync(filePath)) {
    console.error('No processes found.');
    process.exit(1);
  }
  const processes = JSON.parse(await fs.promises.readFile(filePath));
  const pid = processes[name];
  if (!pid) {
    console.error(`No process found with the name "${name}".`);
    process.exit(1);
  }
  console.log(`Killing process ${name}:${pid}...`);
  try {
    process.kill(pid, 'SIGTERM');
  } catch (error) {
    console.error(`Failed to kill process ${name}:${pid}. Continuing cleanup...`);
  }
  const logFile = getLogFile(name);
  await fs.promises.rm(logFile);
  delete processes[name];
  if (Object.keys(processes).length === 0) {
    await fs.promises.rm(filePath);
  } else {
    await fs.promises.writeFile(filePath, JSON.stringify(processes, null, 2));
  }
  process.exit(0);
}

async function retrieveProcess(name) {
  if (!fs.existsSync(filePath)) {
    console.error('No processes found.');
    process.exit(1);
  }
  const processes = JSON.parse(await fs.promises.readFile(filePath));
  const pid = processes[name];
  if (!pid) {
    console.error(`No process found with the name "${name}".`);
    process.exit(1);
  }
  return {
    name,
    pid,
    logFile: getLogFile(name),
  };
}

module.exports = {
  getLogFile,
  retrieveProcess,
  killProcess,
  saveProcess,
};
