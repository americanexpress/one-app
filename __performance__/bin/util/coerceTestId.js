/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs');
const path = require('node:path');
const { diffArrays } = require('diff');
const {
  green, red, dim, underline, strikethrough,
} = require('colorette');
const getResultDir = require('./getResultDir');

function createCoerceTestId(optionName) {
  return function coerceTestId(arg) {
    const dir = /^[./~]/.test(arg) ? path.resolve(process.cwd(), arg) : getResultDir(arg);
    const id = path.basename(dir);

    const dirContents = fs.readdirSync(dir).sort();
    const expectedContents = ['k6-out.json', 'k6-summary.json', 'metrics-k6.json', 'metrics-prom.json'].sort();

    if (!expectedContents.every((file) => dirContents.includes(file))) {
      const contentsDiff = diffArrays(expectedContents, dirContents, { ignoreWhitespace: true });
      const printDiff = contentsDiff.map((part) => (part.added ? strikethrough(green(part.value))
        // eslint-disable-next-line unicorn/no-nested-ternary -- conflicting rules
        : part.removed ? red(part.value) : dim(part.value)));

      throw new Error(`${red(`Bad option ${underline(optionName)}:`)}\n  Invalid test result directory: ${dir}\n\n  ${printDiff.join(', ')}`);
    }

    return { dir, id };
  };
}

module.exports = createCoerceTestId;
