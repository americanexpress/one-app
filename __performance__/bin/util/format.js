/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const { formatRelative, formatDistanceStrict } = require('date-fns');
const { createColors } = require('colorette');

function colors(supportMarkdown = true) {
  const useMarkdown = process.env.MARKDOWN === 'true';
  const colorette = createColors({ useColor: !useMarkdown });
  if (!useMarkdown || !supportMarkdown) return colorette;
  return {
    ...colorette,
    bold: (str) => `**${str}**`,
    italic: (str) => `_${str}_`,
    strikethrough: (str) => `~~${str}~~`,
    underline: (str) => `<u>${str}</u>`,
  };
}

function nth(n) {
  switch (true) {
    case n % 100 === 11:
    case n % 100 === 12:
    case n % 100 === 13:
      return `${n}th`;
    case n % 10 === 1:
      return `${n}st`;
    case n % 10 === 2:
      return `${n}nd`;
    case n % 10 === 3:
      return `${n}rd`;
    default:
      return `${n}th`;
  }
}

function float(n, precision = 3) {
  return Number(n.toFixed(precision)).toString();
}

function append(unit, format = float) {
  return (n) => `${format(n)}${unit}`;
}

function percent(n) {
  return `${float(n * 100)}%`;
}

function bytes(n) {
  if (n === 0) return '0B';
  if (n < 1e3) return `${float(n, 0)}B`;
  if (n < 1e6) return `${float(n / 1e3)}kB`;
  if (n < 1e9) return `${float(n / 1e6)}MB`;
  return `${float(n / 1e9)}GB`;
}

function pValue(p, alpha = 0.05) {
  const { red, bold } = colors();
  if (p < 0.001) return red(bold('*<0.001'));
  if (p <= alpha) return red(`*${float(p)}`);
  return float(p);
}

function seconds(n) {
  if (n === 0) return '0s';
  if (n < 1e-6) return `${float(n * 1e9)}ns`;
  if (n < 1e-3) return `${float(n * 1e6)}Âµs`;
  if (n < 1) return `${float(n * 1e3)}ms`;
  if (n < 60) return `${float(n, 2)}s`;
  if (n < 3600) return `${Math.floor(n / 60)}m ${seconds(n % 60)}`;
  return `${Math.floor(n / 3600)}h ${seconds(n % 3600)}`;
}

function milliseconds(n) {
  return seconds(n / 1e3);
}

function diff(c, v, format) {
  const { green, red, bold } = colors(false);
  const d = c - v;
  if (d === 0) return `=${format(d)}`;
  if (d > 0) return green(`${bold('-')}${format(d)} (${bold('-')}${percent(d / c)})`);
  return red(`${bold('+')}${format(d * -1)} (${bold('+')}${percent((d / c) * -1)})`);
}

function capitalize(string) {
  return string.replace(/^\w/, (c) => c.toUpperCase());
}

function testType(type) {
  const { bold } = colors();
  return `Test type: ${bold(capitalize(type))}`;
}

function time(timestamp, prefix) {
  const { bold } = colors();
  const relative = formatRelative(timestamp, Date.now());
  if (prefix) return `${prefix} ${bold(capitalize(relative))}`;
  return capitalize(relative);
}

function timeDiff(start, end, prefix) {
  const { bold } = colors();
  const formatted = formatDistanceStrict(end, start);
  if (prefix) return `${prefix} ${bold(formatted)}`;
  return formatted;
}

module.exports = {
  nth,
  colors,
  diff,
  float,
  append,
  percent,
  bytes,
  pValue,
  seconds,
  milliseconds,
  capitalize,
  testType,
  time,
  timeDiff,
};
