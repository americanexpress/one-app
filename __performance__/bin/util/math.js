/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

function sum(data) {
  return data.reduce((acc, value) => acc + value, 0);
}
sum.label = 'Total';

function mean(data) {
  return sum(data) / data.length;
}
mean.label = 'Average';

function median(data) {
  const sorted = data.sort((a, b) => a - b);
  return (sorted[Math.floor((data.length - 1) / 2)] + sorted[Math.ceil((data.length - 1) / 2)]) / 2;
}
median.label = 'Median';

function mode(data) {
  const frequencyMap = {};
  let maxFreq = 0;
  let result;
  for (const item of data) {
    frequencyMap[item] = ++frequencyMap[item] || 1;
    if (frequencyMap[item] > maxFreq) {
      maxFreq = frequencyMap[item];
      result = item;
    }
  }
  return result;
}
mode.label = 'Mode';

function min(data) {
  return Math.min(...data);
}
min.label = 'Min';

function max(data) {
  return Math.max(...data);
}
max.label = 'Max';

function percentile(p) {
  return (data) => data.sort((a, b) => a - b)[Math.floor(data.length * (p / 100)) - 1];
}

const p95 = percentile(95);
p95.label = '95th Percentile';

const p90 = percentile(90);
p90.label = '90th Percentile';

module.exports = {
  sum,
  mean,
  median,
  mode,
  min,
  max,
  percentile,
  p95,
  p90,
};
