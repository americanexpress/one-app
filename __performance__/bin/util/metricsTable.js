/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const table = require('markdown-table');
const { Table } = require('console-table-printer');
const k6metrics = require('../k6/metrics');
const promMetrics = require('../prometheus/metrics');

const metrics = { ...k6metrics, ...promMetrics };

function generateMetricsTable({
  data,
  headers,
  processors,
  markdown = false,
  description = false,
  raw = false,
}) {
  const rows = Object.keys(data).map((metric) => [
    metrics[metric].label,
    description && (metrics[metric].description || ' '),
    ...processors.map((processor) => {
      if (processor.length === 1) {
        return raw ? processor(data[metric]) : metrics[metric].format(processor(data[metric]));
      }
      return processor(metric, data[metric], raw ? (x) => x : metrics[metric].format);
    }),
  ].filter(Boolean));

  const tableHeaders = ['Metric', description && 'Description', ...headers].filter(Boolean);

  if (markdown) {
    return table([
      tableHeaders,
      ...rows,
    ]);
  }

  const consoleTable = new Table({
    columns: [
      { name: 'Metric', alignment: 'left', color: 'cyan' },
      description && {
        name: 'Description', alignment: 'left', maxLen: 40, color: 'none',
      },
      ...headers.map((name) => ({ name, alignment: 'left', color: 'none' })),
    ].filter(Boolean),
    colorMap: { none: '' },
  });

  consoleTable.addRows(rows.map((row) => row.reduce((acc, col, i) => ({
    ...acc,
    [tableHeaders[i]]: col,
  }), {})));

  return consoleTable.render();
}

module.exports = generateMetricsTable;
