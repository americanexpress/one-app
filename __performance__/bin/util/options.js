/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const { red, underline, bold } = require('colorette');

const metrics = {
  ...require('../k6/metrics'),
  ...require('../prometheus/metrics'),
};

module.exports.description = {
  alias: 'd',
  description: 'Include a description of each metric in the results table',
  type: 'boolean',
  default: false,
};

module.exports.processors = {
  alias: 'p',
  description: 'processors to run on each column of data',
  type: 'array',
  default: ['mean', 'max'],
  choices: ['sum', 'mean', 'median', 'mode', 'min', 'max', 'p95', 'p90'],
};

module.exports.raw = {
  alias: 'r',
  description: 'Output metrics unformatted (unitless)',
  type: 'boolean',
  default: false,
};

module.exports.metrics = {
  alias: 'm',
  type: 'array',
  description: 'metrics to compare',
  coerce: (arg) => {
    const invalidMetrics = arg.filter((metric) => !Object.keys(metrics).includes(metric));

    if (invalidMetrics.length > 0) {
      throw new Error(`${red(`Bad option: ${underline('metrics')} invalid metric(s) provided`)}\n  Run ${bold('npm run perf -- explain')} to list available metrics.`);
    }

    return arg;
  },
  defaultDescription: `Run ${bold('npm run perf -- explain --defaults')} to list`,
  default: [
    'http_req_duration',
    'http_req_blocked',
    'http_req_waiting',
    'http_req_receiving',
    'nodejs_heap_size_total_bytes',
    'nodejs_heap_size_used_bytes',
    'nodejs_heap_space_size_used_bytes',
    'nodejs_eventloop_lag_seconds',
    'rate(process_cpu_seconds_total[1m])',
    'rate(process_cpu_system_seconds_total[1m])',
    'rate(process_cpu_user_seconds_total[1m])',
    'rate(nodejs_gc_duration_seconds_sum[1m])',
    'histogram_quantile(0.95, sum(rate(nodejs_gc_duration_seconds_bucket[1m])) by (le))',
    'rate(circuit{event="open"}[1m])',
    'rate(circuit{event="close"}[1m])',
  ],
};

module.exports.markdown = {
  description: 'Print the results table in markdown',
  type: 'boolean',
  default: false,
  coerce: (arg) => {
    if (arg === true) {
      process.env.MARKDOWN = 'true';
    }
    return arg;
  },
};
