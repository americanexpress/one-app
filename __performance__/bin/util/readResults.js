/*
 * Copyright 2024 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('node:fs/promises');
const path = require('node:path');

async function readResults(resultsDir) {
  const [k6, prom, summary] = (await Promise.all([
    fs.readFile(path.join(resultsDir, 'metrics-k6.json')),
    fs.readFile(path.join(resultsDir, 'metrics-prom.json')),
    fs.readFile(path.join(resultsDir, 'k6-summary.json')),
  ])).map(JSON.parse);

  return {
    data: { ...k6, ...prom },
    summary,
    get meta() {
      return summary.meta;
    },
  };
}

module.exports = readResults;
