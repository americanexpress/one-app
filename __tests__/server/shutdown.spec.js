/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import util from 'node:util';

describe('shutdown', () => {
  jest.spyOn(global, 'setTimeout').mockImplementation(() => {});
  jest.spyOn(global, 'setImmediate').mockImplementation(() => {});
  jest.spyOn(process, 'exit').mockImplementation(() => {});
  jest.spyOn(console, 'log').mockImplementation(util.format);
  jest.spyOn(console, 'error').mockImplementation(util.format);

  const signalNames = [
    'SIGINT',
    'SIGTERM',
  ];

  function load() {
    jest.resetModules();
    return require('../../src/server/shutdown');
  }

  beforeEach(() => {
    signalNames.forEach((signalName) => process.removeAllListeners(signalName));
  });

  it('shutdown calls the close property of servers added', () => {
    const close = jest.fn();
    const { addServer, shutdown } = load();
    addServer({ close });
    expect(close).not.toHaveBeenCalled();
    shutdown();
    expect(close).toHaveBeenCalledTimes(1);
  });

  it('shutdown logs to the console', () => {
    console.log.mockClear();
    const { shutdown } = load();
    shutdown();
    expect(console.log).toHaveBeenCalled();
    expect(console.log.mock.calls).toMatchSnapshot();
  });

  signalNames.forEach((signalName) => {
    describe(`${signalName}`, () => {
      it('calls shutdown', async () => {
        const close = jest.fn();
        const { addServer } = load();
        addServer({ close });
        expect(close).not.toHaveBeenCalled();
        process.emit(signalName);
        expect(close).toHaveBeenCalledTimes(1);
      });

      it('calls process.exit async when sent twice', () => {
        process.exit.mockClear();
        setImmediate.mockClear();
        load();
        process.emit(signalName);
        expect(process.exit).not.toHaveBeenCalled();
        process.emit(signalName);
        expect(setImmediate).toHaveBeenCalled();
        setImmediate.mock.calls[0][0]();
        expect(process.exit).toHaveBeenCalledTimes(1);
      });

      it('writes an error message when calling process.exit', () => {
        console.error.mockClear();
        load();
        process.emit(signalName);
        expect(console.error).not.toHaveBeenCalled();
        process.emit(signalName);
        expect(setImmediate).toHaveBeenCalled();
        setImmediate.mock.calls[0][0]();
        expect(console.error).toHaveBeenCalled();
        expect(console.error.mock.calls).toMatchSnapshot();
      });

      it('forcibly exits if still running later on', () => {
        // babel-node bug, fixed but not yet published
        process.exit.mockClear();
        setTimeout.mockClear();
        setImmediate.mockClear();
        load();
        process.emit(signalName);
        expect(process.exit).not.toHaveBeenCalled();
        expect(setTimeout).toHaveBeenCalledTimes(1);
        expect(setTimeout.mock.calls[0][1]).toMatchSnapshot();
        setTimeout.mock.calls[0][0]();
        expect(setImmediate).toHaveBeenCalled();
        setImmediate.mock.calls[0][0]();
        expect(process.exit).toHaveBeenCalledTimes(1);
      });

      it('writes an error message if still running later on', () => {
        // babel-node bug, fixed but not yet published
        process.exit.mockClear();
        setTimeout.mockClear();
        console.error.mockClear();
        load();
        process.emit(signalName);
        expect(console.error).not.toHaveBeenCalled();
        expect(setTimeout).toHaveBeenCalledTimes(1);
        setTimeout.mock.calls[0][0]();
        expect(console.error).toHaveBeenCalledTimes(2);
        expect(console.error.mock.results.map((result) => result.value)).toMatchSnapshot();
      });

      it('uses node unref to prevent the exit timeout from preventing exiting', () => {
        process.exit.mockClear();
        setTimeout.mockClear();

        const unref = jest.fn();
        setTimeout.mockImplementationOnce(() => ({ unref }));

        load();
        process.emit(signalName);
        expect(process.exit).not.toHaveBeenCalled();
        expect(setTimeout).toHaveBeenCalledTimes(1);
        expect(unref).toHaveBeenCalledTimes(1);
      });
    });
  });
});
