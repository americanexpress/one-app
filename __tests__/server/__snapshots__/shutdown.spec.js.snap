// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shutdown SIGINT forcibly exits if still running later on 1`] = `10000`;

exports[`shutdown SIGINT writes an error message if still running later on 1`] = `
[
  "still running after 10s, forcibly shutting down",
  "shutting down, forcibly stopping node",
]
`;

exports[`shutdown SIGINT writes an error message when calling process.exit 1`] = `
[
  [
    "shutting down, forcibly stopping node",
  ],
]
`;

exports[`shutdown SIGTERM forcibly exits if still running later on 1`] = `10000`;

exports[`shutdown SIGTERM writes an error message if still running later on 1`] = `
[
  "still running after 10s, forcibly shutting down",
  "shutting down, forcibly stopping node",
]
`;

exports[`shutdown SIGTERM writes an error message when calling process.exit 1`] = `
[
  [
    "shutting down, forcibly stopping node",
  ],
]
`;

exports[`shutdown shutdown logs to the console 1`] = `
[
  [
    "shutting down, closing servers",
  ],
]
`;
